### ISSUE TITLE
static_link_cpp_runtimes doesn't link the libstdc++ statically

### ISSUE THREAD

**Issue by tomgr** [bazelbuild/bazel#14342]
### Description of the problem / feature request:

The `static_link_cpp_runtimes` feature doesn't link libstdc++ statically despite appearing to be intended to as per https://docs.bazel.build/versions/main/cc-toolchain-config-reference.html#well-known-features.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Create the following files:

BUILD:
```
cc_binary(
  name = "test",
  srcs = ["main.cc"],
  features = ["static_link_cpp_runtimes"],
)
```
main.cc:
```
#include <iostream>

int main(int argc, char** argv) { 
  std::cout << "Hello" << std::endl;
}
```
WORKSPACE:
```
```

Running `bazel build //:test && ldd bazel-bin/test` gives:
```
        linux-vdso.so.1 (0x00007fffccffa000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f3b2a2c9000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f3b2a1e5000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3b29fbd000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3b2a4ea000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f3b29fa3000)
```

### What operating system are you running Bazel on?

Ubuntu 22.04, but observed on several other Linuxes

### What's the output of `bazel info release`?

`release 6.0.0-pre.20211110.1`. Using `last_green` with bazelisk also shows the same error (when I ran this last_green was cd4bc7ce3d2a51ecf8df7eea7a8b3b6270357f11).

###  Have you found anything relevant by searching the web?

#8672 has a similar problem with fully_static_link, but this also statically links glibc, which we don't want. The solution there also depends on setting BAZEL_LINKLIBS but we'd like to be able to control which targets link libstdc++ statically individually. static_link_cpp_runtimes is also compiler-independent (if supported by the toolchain), unlike BAZEL_LINKLIBS.

**Comment by keith** [comment]
I just looked briefly but it looks to me like this just isn't supported in bazel, maybe it is in blaze

https://github.com/bazelbuild/bazel/blob/3678d6a84be4508c911e8d7e1d2be9132804f6b3/src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java#L1225

**Comment by tomgr** [comment]
Thanks. I think it's only the auto-configuration that's not enabled as it seems you can define the `static_link_cpp_runtimes` feature yourself. If I copy the auto-generated `local_config_cc` and patch `cc_toolchain_config.bzl` to add:
```bzl
    static_link_cpp_runtimes_feature = feature(
        name = "static_link_cpp_runtimes",
        enabled = False,
        flag_sets = [
            flag_set(
                actions = [
                    ACTION_NAMES.cpp_link_executable,
                    ACTION_NAMES.cpp_link_dynamic_library,
                    ACTION_NAMES.lto_index_for_executable,
                    ACTION_NAMES.lto_index_for_dynamic_library,
                ],
                flag_groups = [flag_group(flags = ["-static-libstdc++"])],
            ),
        ],
    )
```
(and also add this to the list of features that are exported), then I can get static linking to work by setting `features = ["static_link_cpp_runtimes"]`. To ensure that normal cc_binary targets remain dynamically linked I also had to modify the `user_link_flags` feature to add another flag set:
```bzl
            flag_set(
                actions = all_link_actions + lto_index_actions,
                flag_groups = [flag_group(flags = ["-lstdc++"])],
                with_features = [
                    with_feature_set(
                        not_features = ["static_link_cpp_runtimes"],
                    ),
                ],
            ),
```
and then I modified the default value of `BAZEL_LINKLIBS` to remove "-lstdc++".

I wonder if it's worth patching bazel's `unix_cc_configure.bzl` to:
* Accept a user-definable list of flags that would be used to dynamically link the C++ runtime (maybe `BAZEL_CXXSTDLIB_DYNAMIC`). This would default to `-lstdc++` and `-lstdc++` would be removed from `BAZEL_LINKLIBS`.
* Accept a user-definable list of flags that would be used to statically link the C++ runtime (maybe `BAZEL_CXXSTDLIB_STATIC`). This would default to `-static-libstdc++`.

This would make it much easier to statically or dynamically link the c++ standard library than it is now, and would enable users to enable it either per binary, or globally. I'd be happy to work on a patch if this sounds like a useful direction.

**Comment by fmeum** [comment]
I would find this very useful. Making the auto-configured toolchains work well with all advertised Bazel features is also especially helpful for those new to Bazel.

**Comment by fhanau** [comment]
This is an ongoing issue and should not be marked as stale.

The workaround of setting BAZEL_LINKLIBS works for me on Linux (e.g. `BAZEL_LINKLIBS='-l%:libc++.a -lm -static-libgcc'` when using libc++) and by using `--action_env` it can also be used through the command line or bazelrc. Still, it would be much nicer to have the feature flag supported so this can be set on the target level.

**Comment by fmeum** [comment]
@Synss Would you be willing to send a PR to Bazel with this change? It looks good to me and I can also help find a reviewer.

**Comment by Synss** [comment]
Of course! I’ll be happy if the patch gets upstreamed and I don’t have to maintain it! 😉 I’ll look into sending a PR tomorrow.

**Comment by apivovarov** [comment]
Related Discussion:

How to build cpp executable with static libsdtc++.a? -lstdc++ is hardcoded in bazel!!!
https://github.com/bazelbuild/bazel/discussions/23324

Solution:
```
export BAZEL_LINKLIBS="-l%:libstdc++.a:-lm"
```
After that all executables generated by `bazel build` will have libstdc++ linked statically.

**Comment by keith** [comment]
with https://github.com/bazelbuild/rules_cc/commit/5eeedeb388c18cba2855e8936ac387739f9aa03d you can have a lot more flexibility over parts of this
